junit-jupiter-api/src/main/java/org/junit/jupiter/api/AssertThrowsExactly.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2015-2021 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v2.0 which
+ * accompanies this distribution and is available at
+ *
+ * https://www.eclipse.org/legal/epl-v20.html
+ */
+
+package org.junit.jupiter.api;
+
+import static org.junit.jupiter.api.AssertionUtils.buildPrefix;
+import static org.junit.jupiter.api.AssertionUtils.format;
+import static org.junit.jupiter.api.AssertionUtils.getCanonicalName;
+import static org.junit.jupiter.api.AssertionUtils.nullSafeGet;
+
+import java.util.function.Supplier;
+
+import org.junit.jupiter.api.function.Executable;
+import org.junit.platform.commons.util.UnrecoverableExceptions;
+import org.opentest4j.AssertionFailedError;
+
+/**
+ * {@code AssertThrowsExactly} is a collection of utility methods that support asserting
+ * an exception of an expected type is thrown.
+ *
+ * @since 5.8
+ */
+class AssertThrowsExactly {
+
+	private AssertThrowsExactly() {
+		/* no-op */
+	}
+
+	static <T extends Throwable> T assertThrowsExactly(Class<T> expectedType, Executable executable) {
+		return assertThrowsExactly(expectedType, executable, (Object) null);
+	}
+
+	static <T extends Throwable> T assertThrowsExactly(Class<T> expectedType, Executable executable, String message) {
+		return assertThrowsExactly(expectedType, executable, (Object) message);
+	}
+
+	static <T extends Throwable> T assertThrowsExactly(Class<T> expectedType, Executable executable,
+			Supplier<String> messageSupplier) {
+
+		return assertThrowsExactly(expectedType, executable, (Object) messageSupplier);
+	}
+
+	@SuppressWarnings("unchecked")
+	private static <T extends Throwable> T assertThrowsExactly(Class<T> expectedType, Executable executable,
+			Object messageOrSupplier) {
+
+		try {
+			executable.execute();
+		}
+		catch (Throwable actualException) {
+			if (expectedType.equals(actualException.getClass())) {
```

Comment:
Except for this line, this whole method is duplicated and so are most of the tests. Could we remove the duplication here using by moving these methods to `AssertThrows` and a `Predicate<Throwable>` and the test duplication using a common base class with a template method that calls `assertThrows` or `assertThrowsExactly`?

Reply:
Agree. I thought about that. But it seems like not a common practice in a source code of assertions classes, so I split them.

Reply:
Here is an example of removing the production code duplication. What do you think?

```

	static <T extends Throwable> T assertThrows(Class<T> expectedType, Executable executable,
			Supplier<String> messageSupplier) {

		return assertThrows(expectedType, executable, (Object) messageSupplier, Class::isInstance);
	}
	static <T extends Throwable> T assertThrowsExactly(Class<T> expectedType, Executable executable,
			Supplier<String> messageSupplier) {

		return assertThrows(expectedType, executable, (Object) messageSupplier, Class::equals);
	}

	@SuppressWarnings("unchecked")
	private static <T extends Throwable> T assertThrows(Class<T> expectedType, Executable executable,
			Object messageOrSupplier, BiPredicate<Class<T>, Class<?>> exceptionCheck) {

		try {
			executable.execute();
		}
		catch (Throwable actualException) {
			if (exceptionCheck.test(expectedType, actualException.getClass())) {
				return (T) actualException;
			}
			else {
				UnrecoverableExceptions.rethrowIfUnrecoverable(actualException);
				String message = buildPrefix(nullSafeGet(messageOrSupplier))
						+ format(expectedType, actualException.getClass(), "Unexpected exception type thrown");
				throw new AssertionFailedError(message, actualException);
			}
		}

		String message = buildPrefix(nullSafeGet(messageOrSupplier))
				+ String.format("Expected %s to be thrown, but nothing was thrown.", getCanonicalName(expectedType));
		throw new AssertionFailedError(message);
	}
```

I don't know about the test duplication. I understand that those tests are almost completely identical, but their purpose is to test very similar but a bit different assertions. So I think they should be tested separately because they are different assertions and their implementation may vary in the future.

Reply:
I think that looks good! Regarding the tests, if I'm not mistaken there is a subset that's exactly the same except for calling `assertThrowsExactly` instead of `assertThrows`. If so, I'd propose moving those to an abstract super class with an abstract method to call either one in the subclasses and add additional tests for specific behavior there as well.